File: frontend/src/types/finance.ts
export interface GeneralLedgerAccount {
  id: string;
  code: string;
  name: string;
  account_type: 'ASSET' | 'LIABILITY' | 'EQUITY' | 'REVENUE' | 'EXPENSE';
  parent_id?: string;
  opening_balance: number;
  current_balance: number;
  status: 'ACTIVE' | 'INACTIVE';
}

export interface JournalEntryLine {
  id: string;
  account_id: string;
  debit: number;
  credit: number;
  cost_center_id?: string;
  department_id?: string;
  project_id?: string;
  region?: string;
}

export interface JournalEntry {
  id: string;
  entry_number: string;
  entry_date: string;
  description: string;
  status: 'DRAFT' | 'POSTED' | 'VOIDED';
  source_document?: string;
  lines: JournalEntryLine[];
}

export interface InvoiceItem {
  id: string;
  product_id: string;
  quantity: number;
  unit_price: number;
  total_cost: number;
}

export interface CustomerInvoice {
  id: string;
  invoice_number: string;
  customer_id: string;
  sales_order_id?: string;
  issue_date: string;
  due_date: string;
  payment_terms: string;
  subtotal: number;
  tax_amount: number;
  discount_amount: number;
  total_amount: number;
  outstanding_balance: number;
  status: 'DRAFT' | 'PENDING_APPROVAL' | 'APPROVED' | 'ISSUED' | 'PARTIALLY_PAID' | 'PAID' | 'CANCELLED';
  items?: InvoiceItem[];
}

export interface VendorBillItem {
  id: string;
  product_id: string;
  quantity: number;
  unit_price: number;
  total_cost: number;
}

export interface VendorBill {
  id: string;
  bill_number: string;
  supplier_id: string;
  purchase_order_id?: string;
  bill_date: string;
  due_date: string;
  subtotal: number;
  tax_amount: number;
  total_amount: number;
  outstanding_balance: number;
  status: 'DRAFT' | 'APPROVED' | 'PARTIALLY_PAID' | 'PAID' | 'CANCELLED';
  items?: VendorBillItem[];
}

export interface PaymentAllocation {
  id: string;
  payment_id: string;
  invoice_id?: string;
  vendor_bill_id?: string;
  allocated_amount: number;
}

export interface Payment {
  id: string;
  payment_number: string;
  payment_type: 'CUSTOMER_INFLOW' | 'VENDOR_OUTFLOW' | 'REFUND';
  payment_date: string;
  amount: number;
  payment_method: string;
  reference_number?: string;
  status: string;
  allocations: PaymentAllocation[];
}

export interface ExpenseCategory {
  id: string;
  name: string;
  code: string;
  default_account_id?: string;
}

export interface ExpenseClaim {
  id: string;
  claim_number: string;
  employee_id: string;
  expense_category_id: string;
  amount: number;
  claim_date: string;
  description: string;
  receipt_image_url?: string;
  status: 'DRAFT' | 'SUBMITTED' | 'APPROVED' | 'REIMBURSED' | 'REJECTED';
}

export interface Asset {
  id: string;
  asset_number: string;
  name: string;
  category: string;
  purchase_date: string;
  purchase_value: number;
  residual_value: number;
  useful_life_months: number;
  status: 'ACTIVE' | 'DISPOSED' | 'FULLY_DEPRECIATED';
}

export interface BudgetLine {
  id: string;
  account_id: string;
  allocated_amount: number;
  actual_amount: number;
}

export interface Budget {
  id: string;
  name: string;
  cost_center_id: string;
  start_date: string;
  end_date: string;
  status: string;
  lines: BudgetLine[];
}

export interface BankAccount {
  id: string;
  account_number: string;
  bank_name: string;
  account_type: string;
  gl_account_id: string;
  status: string;
}

export interface BankTransaction {
  id: string;
  bank_account_id: string;
  transaction_date: string;
  description: string;
  debit: number;
  credit: number;
  reference_number?: string;
  reconciliation_status: 'MATCHED' | 'PARTIALLY_MATCHED' | 'UNMATCHED' | 'IGNORED';
}

File: backend/app/schemas/finance.py
from datetime import date, datetime
from typing import List, Optional
from uuid import UUID
from pydantic import BaseModel, ConfigDict, Field

class GeneralLedgerAccountBase(BaseModel):
    code: str = Field(..., max_length=50)
    name: str = Field(..., max_length=255)
    account_type: str # ASSET, LIABILITY, EQUITY, REVENUE, EXPENSE
    parent_id: Optional[UUID] = None
    opening_balance: float = 0.0
    status: str = "ACTIVE"

class GeneralLedgerAccountCreate(GeneralLedgerAccountBase):
    pass

class GeneralLedgerAccountResponse(GeneralLedgerAccountBase):
    id: UUID
    current_balance: float
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)

class JournalEntryLineCreate(BaseModel):
    account_id: UUID
    debit: float = 0.0
    credit: float = 0.0
    cost_center_id: Optional[UUID] = None
    department_id: Optional[UUID] = None
    project_id: Optional[UUID] = None
    region: Optional[str] = None

class JournalEntryLineResponse(BaseModel):
    id: UUID
    account_id: UUID
    debit: float
    credit: float
    cost_center_id: Optional[UUID]
    department_id: Optional[UUID]
    project_id: Optional[UUID]
    region: Optional[str]
    model_config = ConfigDict(from_attributes=True)

class JournalEntryCreate(BaseModel):
    entry_date: date
    description: str = Field(..., max_length=512)
    source_document: Optional[str] = None
    lines: List[JournalEntryLineCreate]

class JournalEntryResponse(BaseModel):
    id: UUID
    entry_number: str
    entry_date: date
    description: str
    status: str
    source_document: Optional[str]
    lines: List[JournalEntryLineResponse] = []
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)

class InvoiceItemCreate(BaseModel):
    product_id: UUID
    quantity: float
    unit_price: float

class InvoiceItemResponse(BaseModel):
    id: UUID
    product_id: UUID
    quantity: float
    unit_price: float
    total_cost: float
    model_config = ConfigDict(from_attributes=True)

class CustomerInvoiceCreate(BaseModel):
    customer_id: UUID
    sales_order_id: Optional[UUID] = None
    payment_terms: str = "NET30"
    items: List[InvoiceItemCreate]

class CustomerInvoiceResponse(BaseModel):
    id: UUID
    invoice_number: str
    customer_id: UUID
    sales_order_id: Optional[UUID]
    issue_date: date
    due_date: date
    payment_terms: str
    subtotal: float
    tax_amount: float
    discount_amount: float
    total_amount: float
    outstanding_balance: float
    status: str
    created_at: datetime
    items: List[InvoiceItemResponse] = []
    model_config = ConfigDict(from_attributes=True)

class VendorBillItemCreate(BaseModel):
    product_id: UUID
    quantity: float
    unit_price: float

class VendorBillItemResponse(BaseModel):
    id: UUID
    product_id: UUID
    quantity: float
    unit_price: float
    total_cost: float
    model_config = ConfigDict(from_attributes=True)

class VendorBillCreate(BaseModel):
    supplier_id: UUID
    purchase_order_id: Optional[UUID] = None
    due_date: date
    items: List[VendorBillItemCreate]

class VendorBillResponse(BaseModel):
    id: UUID
    bill_number: str
    supplier_id: UUID
    purchase_order_id: Optional[UUID]
    bill_date: date
    due_date: date
    subtotal: float
    tax_amount: float
    total_amount: float
    outstanding_balance: float
    status: str
    created_at: datetime
    items: List[VendorBillItemResponse] = []
    model_config = ConfigDict(from_attributes=True)

class PaymentAllocationCreate(BaseModel):
    invoice_id: Optional[UUID] = None
    vendor_bill_id: Optional[UUID] = None
    allocated_amount: float

class PaymentAllocationResponse(BaseModel):
    id: UUID
    invoice_id: Optional[UUID]
    vendor_bill_id: Optional[UUID]
    allocated_amount: float
    model_config = ConfigDict(from_attributes=True)

class PaymentCreate(BaseModel):
    payment_type: str # CUSTOMER_INFLOW, VENDOR_OUTFLOW, REFUND
    payment_date: date
    amount: float
    payment_method: str = "CASH"
    reference_number: Optional[str] = None
    allocations: List[PaymentAllocationCreate]

class PaymentResponse(BaseModel):
    id: UUID
    payment_number: str
    payment_type: str
    payment_date: date
    amount: float
    payment_method: str
    reference_number: Optional[str]
    status: str
    created_at: datetime
    allocations: List[PaymentAllocationResponse] = []
    model_config = ConfigDict(from_attributes=True)

class AssetCreate(BaseModel):
    name: str = Field(..., max_length=255)
    category: str = Field(..., max_length=100)
    purchase_date: date
    purchase_value: float
    residual_value: float = 0.0
    useful_life_months: int
    asset_account_id: UUID
    depreciation_account_id: UUID

class AssetResponse(BaseModel):
    id: UUID
    asset_number: str
    name: str
    category: str
    purchase_date: date
    purchase_value: float
    residual_value: float
    useful_life_months: int
    status: str
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)

class BudgetLineCreate(BaseModel):
    account_id: UUID
    allocated_amount: float

class BudgetLineResponse(BaseModel):
    id: UUID
    account_id: UUID
    allocated_amount: float
    actual_amount: float
    model_config = ConfigDict(from_attributes=True)

class BudgetCreate(BaseModel):
    name: str = Field(..., max_length=100)
    cost_center_id: UUID
    start_date: date
    end_date: date
    lines: List[BudgetLineCreate]

class BudgetResponse(BaseModel):
    id: UUID
    name: str
    cost_center_id: UUID
    start_date: date
    end_date: date
    status: str
    lines: List[BudgetLineResponse] = []
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)

class ExpenseClaimCreate(BaseModel):
    expense_category_id: UUID
    amount: float
    claim_date: date
    description: str = Field(..., max_length=512)
    receipt_image_url: Optional[str] = None

class ExpenseClaimResponse(BaseModel):
    id: UUID
    claim_number: str
    employee_id: UUID
    expense_category_id: UUID
    amount: float
    claim_date: date
    description: str
    receipt_image_url: Optional[str]
    status: str
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)
